### ISSUE TITLE
panic with index out of range on slurm18.08.4

### ISSUE THREAD

**Issue by sterni1971** [vpenso/prometheus-slurm-exporter#18]

**Comment by sterni1971** [comment]
Exporter crashes after 10minutes

Slurm 18.08.4

# bin/prometheus-slurm-exporter 
INFO[0000] Starting Server: :8080                        source="main.go:43"
panic: runtime error: index out of range

goroutine 591 [running]:
main.SplitColonValueToFloat(0xc00014b59e, 0x11, 0x0)
        /root/prometheus-slurm-exporter/scheduler.go:61 +0xc9
main.ParseSchedulerMetrics(0xc000260000, 0x10eb, 0x1e00, 0x0)
        /root/prometheus-slurm-exporter/scheduler.go:75 +0x20a
main.SchedulerGetMetrics(0x8f38c8)
        /root/prometheus-slurm-exporter/scheduler.go:83 +0x27
main.(*SchedulerCollector).Collect(0xc00016e240, 0xc00007a3c0)
        /root/prometheus-slurm-exporter/scheduler.go:119 +0x34
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
        /root/prometheus-slurm-exporter/src/github.com/prometheus/client_golang/prometheus/registry.go:442 +0x193
created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
        /root/prometheus-slurm-exporter/src/github.com/prometheus/client_golang/prometheus/registry.go:534 +0xe23
[root@max-adm01 prometheus-slurm-exporter]# bin/prometheus-slurm-exporter

**Comment by erimar77** [comment]
I just had the same crash last evening, but have yet to experience it again after a restart of the service.    It was during a period of general usage without any large spikes of job submission.. only around 1,000 jobs pending at the time.

**Comment by tdthatcher** [comment]
This exporter was written for an older version of slurm. The sdiag-parsing code is a little blunt in that it parses by line number, and sdiag's output changes enough in slurm 18.04 to throw it off. The fix that is working for me so far is to edit scheduler.go with the following and then recompile:

`sm.last_cycle          = SplitColonValueToFloat(lines[20]) // 14->20`
`sm.mean_cycle          = SplitColonValueToFloat(lines[23]) // 17->23`
`sm.cycle_per_minute    = SplitColonValueToFloat(lines[25]) // 19->25`
`sm.backfill_last_cycle = SplitColonValueToFloat(lines[34]) // 27->34`
`sm.backfill_mean_cycle = SplitColonValueToFloat(lines[36]) // 29->36`
`sm.backfill_depth_mean = SplitColonValueToFloat(lines[39]) // 32->39`

I implemented this yesterday in my own install after multiple exporter crashes. So far it's been stable for close to 24 hours, and the resulting data in the matching grafana dash looks much better, too.

**Comment by erimar77** [comment]
For 17.11.7 I had to use:
```
  sm.threads             = SplitColonValueToFloat(lines[4])
  sm.queue_size          = SplitColonValueToFloat(lines[5])
  sm.last_cycle          = SplitColonValueToFloat(lines[18])
  sm.mean_cycle          = SplitColonValueToFloat(lines[21])
  sm.cycle_per_minute    = SplitColonValueToFloat(lines[23])
  sm.backfill_last_cycle = SplitColonValueToFloat(lines[32])
  sm.backfill_mean_cycle = SplitColonValueToFloat(lines[34])
  sm.backfill_depth_mean = SplitColonValueToFloat(lines[37])
```

**Comment by mtds** [comment]
Hi all.

Yes, as @tdthatcher wrote, this exporter is using a very straightforward way to parse the output
of ``sdiag``. Which is unfortunately very dependent on the version of Slurm: this bug is very similar
to one of the first I have fixed by myself and now it happened again :-(

It would be nice to come up with a smarter way of parsing the output of Slurm cmd line utilities but
at the moment we are busy with other stuff so it's unlikely we will have time to rewrite the parsing
functions.

I am unsure how to fix this bug unless we decide that this exporter is fully functional only
for the latest (say) two versions of Slurm and for older versions you will have to use a previous
release.

I will think about it the next days and leave this bug report open for the moment.

**Comment by erimar77** [comment]
Interestingly enough, it crashed again this evening at the exact same time 24 hours later.
```
Jun 24 19:00:02 slurm_exporter[23083]: panic: runtime error: index out of range
Jun 24 19:00:02 slurm_exporter[23083]: goroutine 9784 [running]:
Jun 24 19:00:02 slurm_exporter[23083]: main.SplitColonValueToFloat(0xc42028c243, 0x11, 0x403c000000000000)
Jun 24 19:00:02 slurm_exporter[23083]: #011/root/prometheus-slurm-exporter/scheduler.go:61 +0xcc
Jun 24 19:00:02 slurm_exporter[23083]: main.ParseSchedulerMetrics(0xc420378000, 0x3370, 0x3e00, 0x7f525b312e58)
Jun 24 19:00:02 slurm_exporter[23083]: #011/root/prometheus-slurm-exporter/scheduler.go:74 +0x1ca
Jun 24 19:00:02 slurm_exporter[23083]: main.SchedulerGetMetrics(0xc42002e500)
Jun 24 19:00:02 slurm_exporter[23083]: #011/root/prometheus-slurm-exporter/scheduler.go:83 +0x27
Jun 24 19:00:02 slurm_exporter[23083]: main.(*SchedulerCollector).Collect(0xc4200d25c0, 0xc42078e240)
Jun 24 19:00:02 slurm_exporter[23083]: #011/root/prometheus-slurm-exporter/scheduler.go:119 +0x34
Jun 24 19:00:02 slurm_exporter[23083]: github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
Jun 24 19:00:02 slurm_exporter[23083]: #011/root/prometheus-slurm-exporter/src/github.com/prometheus/client_golang/prometheus/registry.go:442 +0x199
Jun 24 19:00:02 slurm_exporter[23083]: created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
Jun 24 19:00:02 slurm_exporter[23083]: #011/root/prometheus-slurm-exporter/src/github.com/prometheus/client_golang/prometheus/registry.go:534 +0xe90

Jun 25 19:00:02 slurm_exporter[28617]: panic: runtime error: index out of range
Jun 25 19:00:02 slurm_exporter[28617]: goroutine 50204 [running]:
Jun 25 19:00:02 slurm_exporter[28617]: main.SplitColonValueToFloat(0xc420100244, 0x11, 0x4035000000000000)
Jun 25 19:00:02 slurm_exporter[28617]: #011/root/prometheus-slurm-exporter/scheduler.go:61 +0xcc
Jun 25 19:00:02 slurm_exporter[28617]: main.ParseSchedulerMetrics(0xc4201c6000, 0x33bf, 0x3e00, 0x60)
Jun 25 19:00:02 slurm_exporter[28617]: #011/root/prometheus-slurm-exporter/scheduler.go:74 +0x1ca
Jun 25 19:00:02 slurm_exporter[28617]: main.SchedulerGetMetrics(0x100c420558030)
Jun 25 19:00:02 slurm_exporter[28617]: #011/root/prometheus-slurm-exporter/scheduler.go:83 +0x27
Jun 25 19:00:02 slurm_exporter[28617]: main.(*SchedulerCollector).Collect(0xc4200d2580, 0xc420172420)
Jun 25 19:00:02 slurm_exporter[28617]: #011/root/prometheus-slurm-exporter/scheduler.go:119 +0x34
Jun 25 19:00:02 slurm_exporter[28617]: github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
Jun 25 19:00:02 slurm_exporter[28617]: #011/root/prometheus-slurm-exporter/src/github.com/prometheus/client_golang/prometheus/registry.go:442 +0x199
Jun 25 19:00:02 slurm_exporter[28617]: created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
Jun 25 19:00:02 slurm_exporter[28617]: #011/root/prometheus-slurm-exporter/src/github.com/prometheus/client_golang/prometheus/registry.go:534 +0xe90
Jun 25 19:00:02 systemd[1]: slurm_exporter.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Jun 25 19:00:02 systemd[1]: slurm_exporter.service: Failed with result 'exit-code'.
```

**Comment by tdthatcher** [comment]
Ah, hmm, I actually just had the same crash last night, even after fixing the line numbers to make sure the right values are being read. It was up for about 24 hours, as well, instead of crashing multiple times a day as it was before.

**Comment by erimar77** [comment]
Looks like it occurs when Slurm rolls over its sdiag output for the day, note the `Data since` time. 

```
~$ sdiag | head
*******************************************************
sdiag output at Wed Jun 26 13:32:24 2019 (1561573944)
Data since      Tue Jun 25 19:00:00 2019 (1561507200)
*******************************************************
Server thread count:  3
Agent queue size:     0
```

**Comment by mtds** [comment]
Quick update: I have just pushed a modified version of ``scheduler.go`` into the **regexp_parse** branch. I have quickly tested it with ``scheduler_test.go`` and it seems to work correctly.

I can quickly setup a cluster with VMs but it would be better if all of you will test this modification 
on a real deployment (our cluster is still running a stone age old version of Slurm so it's not a
good candidate).

The new method to extract info from the output of ``sdiag`` should work no matter how the output on the console changes. It's certainly not the best Golang code around but it should do the trick.

Annoyingly enough, most of the Slurm cmd line utilities emit output which is definitely not 'machine friendly': it's nice for humans to check but on bigger installations you'll definitely want to export the
data in different ways.

**Comment by tdthatcher** [comment]
Nice, thanks for the update. Giving it a test today.

**Comment by sterni1971** [comment]
Great! Started the test

**Comment by sterni1971** [comment]
Everything okay. IMHO fixed

**Comment by mtds** [comment]
> Everything okay. IMHO fixed

Ok, I will wait till I get feedback from at least one other user in this thread, then I will
push the modified version into the master branch (most likely it will be in the next week 
directly).

**Comment by mtds** [comment]
Hi all.

If your tests are completed I believe I will merge the new ``scheduler.go`` into the master branch tomorrow and then close this issue. If something comes up again we can still reopen it.

**Comment by tdthatcher** [comment]
So far it's holding up for me very well, too. Thanks a lot for the fix!

**Comment by mtds** [comment]
Fixed (see commit f05c8e7).
